sqlmesh/core/macros.py

```diff
@@ -1100,6 +1100,65 @@ def var(
     return exp.convert(evaluator.var(var_name.this, default))
 
 
+@macro()
+def deduplicate(
+    evaluator: MacroEvaluator,
+    relation: exp.Expression,
+    partition_by: t.List[exp.Expression],
+    order_by: t.List[str],
+) -> exp.Query:
+    """Returns a QUERY to deduplicate rows within a table
+
+    Args:
+        relation: table or CTE name to deduplicate
+        partition_by: column names, or expressions to use to identify a window of rows out of which to select one as the deduplicated row
+        order_by: A list of strings representing the ORDER BY clause
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlglot.schema import MappingSchema
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "@deduplicate(demo.table, [user_id, cast(timestamp as date)], ['timestamp desc', 'status asc'])"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT * FROM demo.table QUALIFY ROW_NUMBER() OVER (PARTITION BY user_id, CAST(timestamp AS DATE) ORDER BY timestamp DESC, status ASC) = 1'
+    """
+    if not isinstance(partition_by, list):
+        raise SQLMeshError(
+            "partition_by must be a list of columns: [<column>, cast(<column> as <type>)]"
+        )
+
+    if not isinstance(order_by, list):
+        raise SQLMeshError("order_by must be a list of strings: ['<column> <asc|desc>']")
+
+    partition_clause = exp.tuple_(
+        *[
+            col
+            if not isinstance(col, exp.Cast)
+            else exp.func("cast", this=col.this, to=col.args.get("to"))
+            for col in partition_by
+        ]
+    )
+
+    order_expressions = [
+        evaluator.parse_one(order_item, into=exp.Ordered) for order_item in order_by
```

Comment:
I think we need to set dialect here because if there's no `ASC` or `DESC`, we'll use the wrong default. Different dialects specify different default orders (nulls first/last is also affected)

Reply:
will do

Reply:
updated the error message